File: spring-mvc/src/main/java/info/novatec/footballmanager/FootballerRestController.java
package info.novatec.footballmanager;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

@RequestMapping("/footballers")
@RestController
public class FootballerRestController {

  @Autowired private FootballerRepository footballerRepository;

  @GetMapping
  public List<Footballer> search(
      @RequestParam(name = "position", required = false) String position) {
    if (position == null) {
      return footballerRepository.findAll();
    } else {
      return footballerRepository.findByPosition(position);
    }
  }

  @GetMapping("/{id}")
  public ResponseEntity<Footballer> get(@PathVariable Long id) {
    Optional<Footballer> footballer = footballerRepository.findById(id);
    return footballer.map(ResponseEntity::ok).orElseGet(() -> ResponseEntity.notFound().build());
  }

  @Transactional
  @PostMapping
  public ResponseEntity<Footballer> create(@RequestBody Footballer footballer) {
    return ResponseEntity.status(HttpStatus.CREATED).body(footballerRepository.save(footballer));
  }

  @Transactional
  @DeleteMapping("/{id}")
  public ResponseEntity<Footballer> delete(@PathVariable Long id) {
    Optional<Footballer> footballer = footballerRepository.findById(id);
    footballer.ifPresent(value -> footballerRepository.delete(value));
    return ResponseEntity.noContent().build();
  }
}

File: micronaut/src/main/kotlin/info/novatec/FootballerRestController.kt
package info.novatec

import io.micronaut.http.HttpResponse
import io.micronaut.http.annotation.*
import java.util.*
import javax.transaction.Transactional

@Controller("/footballers")
open class FootballerRestController(
        private val repository: FootballerRepository
) {

    @Get("/")
    fun search(@QueryValue("position") position: String?): List<Footballer> {
        return if (position != null) {
            repository.findByPosition(position)
        } else {
            repository.findAll().toList()
        }
    }

    @Get("/{id}")
    fun get(id: Long): Optional<Footballer> {
        return repository.findById(id)
    }

    @Post("/")
    @Transactional
    open fun create(@Body footballer: Footballer): HttpResponse<Footballer> {
        return HttpResponse.created(repository.save(footballer))
    }

    @Delete("/{id}")
    @Transactional
    open fun delete(id: Long): HttpResponse<Footballer> {
        repository.deleteById(id)
        return HttpResponse.noContent<Footballer>()
    }
}